### DISCUSSION TITLE
How to prevent queue deletion when queue is not empty?(RabbitMQ Cluster Operator)

### DISCUSSION THREAD

**Discussion by Binghankina** [rabbitmq/rabbitmq-server#7737] [Questions]
Hi,
we want to block the deletion of a queue if it's not empty. Currently, we use the messaging topology operator to create a queue and if someone just deletes the queue CR, then the queue is just deleted, even if it is not empty. We are thinking to implement a webhook to prevent the delete request when the queues are not empty with the admission controller. However, do you think that there could be a better way to achieve it?

**Comment by michaelklishin** [comment]
The `queue.delete` operation has a conditions `if_empty` flag. The K8S Operator could be made optionally specify it (I don't think it is possible right now).

**Reply by Binghankina** [reply]
Hi Michael,

Thanks for the answer. We can extend operator to prevent queue delete with condition if_empty. However, even we keep the queue, the CR will be deleted anyway. So when the operator is installed again, it will be state mismatch, isn't it? Do you think the webhook solution is better?

BRs,
Binghan

**Reply by MirahImage** [reply]
I think an admission controller webhook is a good option for preventing deletion of non-empty queues, especially as this policy wouldn't necessarily be universal—some users may wish to delete non-empty queues, which is why this is configurable within RabbitMQ.

**Reply by Zerpet** [reply]
I'm not a fan of the webhook solution because the webhook will have to query external state in order to make a decision, such query takes time, may even time out and never complete, making the webhook relatively slow. Slow webhooks are a terrible experience, in the form of `kubectl` "hanging" without an evident reason why, and it's hard to observe what's going on (i.e. slow webhook).

That being said, there's no other way to prevent a CR to be marked as `Terminating` other than with an admission controller. However, I would not unconditionally prevent deletion of non-empty queues, but extend the `Queue` CR to expose the behaviour that Michael mentions.
